osu.Game/Input/ActionMappingInputManager.cs

```diff
@@ -1,74 +1,133 @@
+                    mappings.Add(b);
+            }
         }
 
-        private void mapKey(InputState state, Key key)
+        private readonly List<Binding> pressedBindings = new List<Binding>();
+
+        protected override bool PropagateKeyDown(IEnumerable<Drawable> drawables, InputState state, KeyDownEventArgs args)
         {
-            T mappedData;
-            if (Mappings.TryGetValue(key, out mappedData))
-                state.Data = mappedData;
+            bool handled = false;
+
+            if (!args.Repeat && (concurrencyMode > ConcurrentActionMode.None || pressedBindings.Count == 0))
+            {
+                Binding validBinding;
+
+                while ((validBinding = mappings.Except(pressedBindings).LastOrDefault(m => m.Keys.CheckValid(state.Keyboard.Keys, concurrencyMode == ConcurrentActionMode.None))) != null)
+                {
+                    if (concurrencyMode == ConcurrentActionMode.UniqueAndSameActions || pressedBindings.All(p => p.Action != validBinding.Action))
+                        handled = drawables.OfType<IHandleActions<T>>().Any(d => d.OnPressed(validBinding.GetAction<T>()));
+
+                    // store both the pressed combination and the resulting action, just in case the assignments change while we are actuated.
+                    pressedBindings.Add(validBinding);
+                }
+            }
+
+            return handled || base.PropagateKeyDown(drawables, state, args);
         }
 
-        private T parseStringRepresentation(string str)
+        protected override bool PropagateKeyUp(IEnumerable<Drawable> drawables, InputState state, KeyUpEventArgs args)
         {
-            T res;
+            bool handled = false;
+
+            foreach (var binding in pressedBindings.ToList())
+            {
+                if (!binding.Keys.CheckValid(state.Keyboard.Keys, concurrencyMode == ConcurrentActionMode.None))
+                {
+                    // clear the no-longer-valid combination/action.
+                    pressedBindings.Remove(binding);
 
-            if (Enum.TryParse(str, out res))
-                return res;
+                    if (concurrencyMode == ConcurrentActionMode.UniqueAndSameActions || pressedBindings.All(p => p.Action != binding.Action))
+                    {
+                        // set data as KeyUp if we're all done with this action.
+                        handled = drawables.OfType<IHandleActions<T>>().Any(d => d.OnReleased(binding.GetAction<T>()));
+                    }
+                }
+            }
 
-            return default(T);
+            return handled || base.PropagateKeyUp(drawables, state, args);
         }
     }
+
+    public enum ConcurrentActionMode
+    {
+        /// <summary>
+        /// One action can be pressed at once. The first action matching a chord will take precedence and no other action will be pressed until it has first been released.
+        /// </summary>
+        None,

```

Comment:
Consider changing the behaviour of modifier keys to match OSes closer (modifiers shouldn't be checked until a non-modifier is pressed).